### PR TITLE
Modify dynamic_quantize ref and opt kernel

### PR DESCRIPTION
**Pull request by byungilm**
+ Bugfix for values only on Y axis
+ Removed hard-coded size

### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl [hunk 1]
```diff
@@ -8,6 +8,14 @@
 #error "dynamic_quantize_gpu_opt.cl: Unsupported output dimension"
 #endif
 
+#define VLOAD_N CAT(vload, VEC_SIZE)
+#define VSTORE_N CAT(vstore, VEC_SIZE)
+#define HALF_N CAT(half, VEC_SIZE)
+#define CONVERT_CHAR_N CAT(convert_char, VEC_SIZE)
+#define AS_TYPE_N_(type, n, x) as_##type##n(x)
+#define AS_TYPE_N(type, n, x) AS_TYPE_N_(half, n, x)
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: why is it fixed as `half`?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl [hunk 2]
```diff
@@ -8,6 +8,14 @@
 #error "dynamic_quantize_gpu_opt.cl: Unsupported output dimension"
 #endif
 
+#define VLOAD_N CAT(vload, VEC_SIZE)
+#define VSTORE_N CAT(vstore, VEC_SIZE)
+#define HALF_N CAT(half, VEC_SIZE)
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: what about using an existing macro such as `MAKE_VECTOR_TYPE(INPUT0_TYPE, VEC_SIZE)`?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl [hunk 3]
```diff
@@ -8,6 +8,20 @@
...
+    #define AS_HALF_TYPE_N(x) (half)
+#else
+    #define VLOAD_N CAT(vload, VEC_SIZE)
+    #define VSTORE_N CAT(vstore, VEC_SIZE)
+    #define HALF_N CAT(half, VEC_SIZE)
+    #define CONVERT_CHAR_N CAT(convert_char, VEC_SIZE)
+    #define AS_TYPE_N_(type, n, x) as_##type##n(x)
+    #define AS_TYPE_N(type, n, x) AS_TYPE_N_(half, n, x)
+    #define AS_HALF_TYPE_N(x) AS_TYPE_N(half, VEC_SIZE, x)
+#endif
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl]: do we need these changes? did original implementation have issue with unaligned case?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/dynamic_quantize/dynamic_quantize_kernel_opt.cpp [hunk 4]
```diff
@@ -6,7 +6,22 @@
...
+static size_t get_match_vector_size(const dynamic_quantize_params& params) {
+    auto block_sizes = { 8, 4, 2 };
+
+    for (auto block_size : block_sizes) {
+        if (((params.inputs[0].X().v * params.inputs[0].Y().v) / simd) % block_size == 0) {
+            return block_size;
+        }
+    }
+
+    return 1;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/dynamic_quantize/dynamic_quantize_kernel_opt.cpp:34]: does opt kernel works with `vec_size=1`?

Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/dynamic_quantize/dynamic_quantize_kernel_opt.cpp:34]: Now validate functions returns false if vec_size is 1.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl (hunk 3 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl]: original code just does vload8 without checking size even x axis is unaligned or 1.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl:16]: Original code was implemented as half8. I thought the target was half. Do we need to consider all format? I will fix it by coming commit.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl (hunk 2 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: I will update it.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: Fixed to INPUT0_TYPE

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl (hunk 2 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: Fixed to MAKE_VECTOR_TYPE(INPUT0_TYPE, VEC_SIZE)

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl (hunk 3 above)
Comment (isanghao) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_ref.cl]: It calls vload8 with INPUT0_SIZE_X / 8. Isn't it sufficient?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl [hunk 5]
```diff
@@ -8,39 +8,95 @@
...
+#define AS_TYPE_N(type, n, x) AS_TYPE_N_(type, n, x)
+#define AS_INPUT_TYPE_N(x) AS_TYPE_N(INPUT0_TYPE, VEC_SIZE, x)
+
 KERNEL(dynamic_quantize_gpu_opt)(
     OPTIONAL_SHAPE_INFO_ARG
     const __global INPUT0_TYPE* input,
     __global OUTPUT_TYPE* output,
     __global OUTPUT1_TYPE* output_scale)
 {
+#if 0
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: Are you going to remove this block? Then please remove it.

Comment (byungilm) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/dynamic_quantize_gpu_opt.cl]: Removed.

**Comment by byungilm** [comment]
Reverted fix for dynamic_quantize_ref by request Mingyu

**Review by isanghao — APPROVED** [review]

**Comment by isanghao** [comment]
I merged this manually because it has conflict with latest branch. Let me close this PR.

**Pull request closed without merging**